lib/record_store/provider/ns1.rb

```diff
@@ -219,6 +227,23 @@ def symbolize_keys(hash)
       def secrets
         super.fetch('ns1')
       end
+
+      def api_rate_limit
+        uri = URI('https://api.nsone.net/v1/monitoring/regions')
```

Comment:
Why are we doing that extra call before any other API call? That's not necessarily the endpoint all the other calls are going to be using. IIRC, in `ns1-go`, they were using the headers of the call they just made, weren't they?
In this case, you are:
1/ duplicating every API call by adding that GET call (hence actively reducing the rate at which we can effectively use the API)
2/ considering only the headers giving us the token information for the GET call (as discussed with NS1: GET, PUT, DELETE and POST have all different token counters, so here we would only know about the GET tokens no matter what method the call used).

I realize that `client.record` might not return the headers at this point, which means we might need to patch the method that actually does the API call to either return the headers or implement the rate limiting approach!

Reply:
It might be dumb question but, aren't they duplicating calls anyways if I call them from other URIs?
- `https://api.nsone.net` for zones and records
- `https://api.nsone.net/v1/monitoring/regions` for headers (nothing related to zones and records)

Reply:
Third option might take longer than expected. I just checked the closed and opened PR in the repo. Last closed PR was merged Dec, 2018, opened PR is still waiting for approval from Jul, 2019.

Reply:
I will probably go with ns1-ruby patch.
>lib/ns1/api/regions.rb
```ruby
module NS1
  module API
    module Regions
      #
      # Returns the list of all monitoring regions for the account
      #
      # @return [NS1::Response]
      #
      def regions()
        perform_request(HTTP_GET, "/v1/monitoring/regions")
      end
    end
  end
end
```
>lib/ns1/api.rb
```ruby
    def self.included(base)
      base.send :include,
        NS1::API::Zones,
        NS1::API::Records
        NS1::API::Records,
        NS1::API::Regions
    end
```
Something like that..

Reply:
It's unfortunate that [only the body is captured](https://github.com/kitop/ns1-ruby/blob/2635616453b37b4ab7e417c08f401ec4e694622c/lib/ns1/transport/net_http.rb#L28-L38) in `ns1` 😢 

I agree with @XaF that it's weird that we're making calls outside of the `ns1` gem...and it's also weird that we're checking the headers on a request that's unrelated to the request we're making (this will result in our _usually_ halving the number of tickets we have, but _usually_...we've established that there are different buckets of tickets for POST, GET, etc).

He's spot on on options here.

Reply:
From @cttttt 's comment and your analysis of their merging of the requests, patching (from this project) the `NS1::Transport::process_response` method might be a good in-between for this, as this methods has the `response` and thus the headers! (this is option 1 above) This would be patching the client even though we're not having the upstream client profit from that patch!
Option 2 of writing fully our own client could work too, but I feel like it would take longer. Up to you to decide at this point @gooallen !

Reply:
Yeah, it's very unfortunate.. :( 
Seems like the option 1 is the best choice for now than the option 2 taking longer. Let me check `NS1::Transport::process_response` method as @XaF suggested to take a look. Let me see how it goes. Thanks, both!

Reply:
It seems possible to call headers from ns1-ruby. Hopefully, it works out in the way we want :D